monix-tail/shared/src/main/scala/monix/tail/internal/IterantMapBatch.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2014-2018 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.tail.internal
+
+import cats.effect.Sync
+import cats.syntax.all._
+import monix.execution.misc.NonFatal
+import monix.tail.Iterant
+import monix.tail.Iterant.{Halt, Last, Next, NextBatch, NextCursor, Suspend}
+import monix.tail.batches.Batch
+import monix.tail.internal.IterantUtils.signalError
+
+import scala.collection.mutable.ArrayBuffer
+
+private[tail] object IterantMapBatch {
+  /**
+    * Implementation for `Iterant#mapBatch`
+    */
+  def apply[F[_], A, B](source: Iterant[F, A], f: A => Batch[B])
+    (implicit F: Sync[F]): Iterant[F, B] = {
+
+    def processSeq(ref: NextCursor[F, A]): NextBatch[F, B] = {
+      val NextCursor(cursor, rest, stop) = ref
+      val buffer = ArrayBuffer.empty[B]
+      var toProcess = cursor.recommendedBatchSize
+      var nextRef: F[Iterant[F, B]] = null.asInstanceOf[F[Iterant[F, B]]]
+
+      // protects against infinite cursors
+      while (toProcess > 0 && cursor.hasNext()) {
+        val cursorB = f(cursor.next()).cursor()
+        while (toProcess > 0 && cursorB.hasNext()) {
```

Comment:
When that mapping function gives you a `Batch`, there's no point in splitting it any further, because the `Batch` is already built for you and you can just serve it as is in a `NextBatch` node. All operations need to handle fairness concerns via `recommendedBatchSize`, so even if you serve bigger batches via this function, it doesn't matter.

The point of `mapBatch` is to be used in certain cases for optimizing a stream, e.g. for turning an `Iterant[F, Array[A]]` into an `Iterant[F, A]`. If we end up splitting those batches, then it becomes too heavy. At that point we might as well work with `flatMap` directly:

```scala
def mapBatch[F[_] : Sync, A](fa: Iterant[F, A])(f: A => Batch[B]): Iterant[F, B] =
  fa.flatMap { a =>
    Iterant.nextBatchS(f(a), Iterant.empty, F.unit)
  }
```

So here what I see happening is no batched processing at all, e.g..

```scala
val batch = f(cursor.next()).cursor()
val next = 
  if (cursor.hasNext()) NextCursor(cursor, rest.map(loop), stop)
  else rest.map(loop)

NextBatch(batch, next, stop)
```